docs/design_proposals/sync-auto-support.md

```diff
@@ -0,0 +1,292 @@
+      dest: "app/classes",
+    },
+    {
+      src: “target/resources"
+      dest: "app/resources",
+    }
+  ]
+  "direct": [
+    {
+      src: "src/main/extra1",
+      dest: "/"
+    },
+    {
+      src: "src/main/extra2",
+      dest: "/"
+    },
+    {
+      src: ".m2/some/dep/my-dep-SNAPSHOT.jar",
+      dest: "app/dependencies/my-dep-SNAPSHOT.jar"
+    }
+  }
+}
+```
+
+Files in the `generated` section will trigger a partial rebuild of the container 
+(everything before containerization) while files in the `direct` section can
+just be synchronized to the running container without a rebuild of anything.
+
+##### Sync or Rebuild?
+
+Each builder implementing an `auto` builder should be able to decide when a sync
+should be skipped and a rebuild done instead. In the jib case for instance, a
+rebuild will be triggered if:
+- a build file has changed (`build.gradle`, `pom.xml`, etc)
+- a file is deleted
+
+#### Open Issues/Questions
+
+**What about files that have specific permissions?**
+
+Jib allows users to customize file permissions, for instance a file on the
+container can be configured to be 0x456 or something. One option instead of
+dealing with this, is to just make all sync'd files 777? Or we allow passthrough
+of permissions from the build system to the sync system.
+
+**Should we allow the user to configure the auto block?**
+
+Perhaps the user knows something that jib doesnt, and wants to ignore some files
+from synchronization. They might want to do:
+
+```
+sync:
+  auto:
+    ignored:
+    - "src/main/jib/myExtraFilesThatBreakADeployment"
+```
+
+TODO (comment here if you need this be elaborated more on)
+
+## Implementation plan
+
+- [`schemas/<version>.go`] Add `Auto` to the schema under `Sync`
+- Before first build, initialize the `sync` state for builders in `auto` mode, this
+sync state is saved in the builder's specific implementation of `auto`
+
+- On file change in dev mode: 
+```
+if (files were deleted)
+  return REBUILD
+
+if (changes were made to build def)
+  return REBUILD
+
+lastSyncState = syncStates["this project"]
+
+if (if all files changes are in lastSyncState.direct)
+  return SYNC{list of direct files}
+
+newSyncState = buildAndCalculateNewSyncState("this project")
+syncStates["this project"] = newSyncState
+
+diff = diff(lastSyncState, newSyncState)
+return SYNC{files in diff}
+```
+
+## Integration test plan
+
+*TBD*
+
+## Alternatives Explored
+
+The following implementation had a few problems and that were potential deal
+breakers, this implementation is left in here for information purposes, no
+attempt was made to implement this. It has the following issues:
+1. Depends on `sync` using `tar` which should be an implementation detail
+2. Doesn't really provide a good base for exposing the `auto` sync configuration
+   block. As skaffold takes care of fewer things, the user must be responsible
+   for implementing it.
+
+#### Delegate generation of sync tar to builder
+
```

Comment:
This works for solutions where we can influence the development of the build system (i.e. jib). But it can't be the only solution, as if we can't influence the build system, we need to implement the solution in skaffold.